### DISCUSSION TITLE
Odd error when I query() using the Java API

### DISCUSSION THREAD

**Discussion by fgreco55** [milvus-io/milvus#27388] [Q&A and General discussion] [answered]
I'm getting an odd error using the Java API when querying.  The server tells me I have 100 rows in my Collection.  But when I call query(), I get:

ERROR: QueryRequest failed:failed to query: attempt #0: fail to get shard leaders from QueryCoord: collection=444516323785331261: collection not loaded: unrecoverable error

I'm not sure what this error is telling me.  What is a "shard leader" and what is a "QueryCoord"?  I must be doing something fundamentally wrong, but I can't find info on this error in the docs.

Here's a code segment:

        mc.loadCollection(
                LoadCollectionParam.newBuilder()
                        .withCollectionName(coll)
                        .build()
        );
        List<String> query_output_fields = Arrays.asList(FIELD1, FIELD2);

        QueryParam queryParam = QueryParam.newBuilder()
                .withCollectionName(coll)
                .withConsistencyLevel(ConsistencyLevelEnum.STRONG)
                .withExpr(query)      // query string goes here
                .withOutFields(query_output_fields)
                .withOffset(0L)
                .withLimit(10L)         // max entries returned
                .build();
        R<QueryResults> respQuery = mc.query(queryParam);                            //  this line gives me the error

**Comment by yhmo** [comment]
"collection not loaded: unrecoverable error" ----- this error is returned from server side. 

Did you check the response of loadCollection()?

```
R<RpcStatus> loadR = client.loadCollection(LoadCollectionParam.newBuilder()
                .withCollectionName(randomCollectionName)
                .build());
if (r.getStatus() != R.Status.Success.getCode()) {
    System.out.print(r.getMessage());
}
```

**Reply by fgreco55** [reply]
Thanks for the suggestion and reply.  Truly thankful the team is answering questions here.

I do get an error from the loadCollection(), but the error is:

       _there is no vector index on collection: _foo_, please create index firstly_

But I do have an index.  After I insert data, I do this:

CreateIndexParam indexParam = CreateIndexParam.newBuilder()
                .withCollectionName(coll)
                .withFieldName(FIELD2)
                .withIndexType(IndexType.TRIE)
                .withMetricType(MetricType.L2)
                .withSyncMode(Boolean.TRUE)
                .build();
        R<RpcStatus> createIndexR = mc.createIndex(indexParam);

        if (createIndexR.getStatus() != R.Status.Success.getCode()) {
            System.out.print("***ERROR:  " + createIndexR.getMessage());
        } else {
            System.out.println("Success creating the INDEX...");                        /* I get Success when I create the index */
        }

So why would I get the "no vector index" error?    Actually, why would I even need an index for a tiny helloworld dataset (100 rows with an int64+String+2DFloat vector?

**Reply by fgreco55** [reply]
I'm good now.

Here's what I learned:

The index is unlike a RDBMS index; it must be associated with the vector field.  You must have an index on the vector field despite what the docs say (it says optional for small datasets).

My query() worked.  Now to move to search().

Thank you for your support.  This was very helpful.  I plan on presenting on my findings at an upcoming conference.

**Reply by fgreco55** [reply, marked as answer]
The docs say this:  "Vector indexes are an organizational unit of metadata used to accelerate [vector similarity search](https://milvus.io/docs/search.md). Without the index built on vectors, Milvus will perform a brute-force search."

This isn't true.  Apparently, it is mandatory to create an index on your vector field.

**Reply by xiaofan-luan** [reply]
Thanks for pointing out. you need to create FLAT index

**Comment by yhmo** [comment]
IndexType.FLAT is brute-force search.
```
CreateIndexParam indexParam = CreateIndexParam.newBuilder()
.withCollectionName(coll)
.withFieldName(xxx)
.withIndexType(IndexType.FLAT)
.build();
```

We just want to unify the workflow.
The following workflow are valid:
1. createCollection ---> createIndex ---> load ---> insert ---> search/query             (mainly for streaming data)
2. createCollection ---> load ---> createIndex ---> insert ---> search/query             (mainly for streaming data)
3. createCollection ---> insert ---> createIndex ---> load ---> search/query             (mainly for static data)
4. createCollection ---> createIndex ---> insert ---> load ---> search/query             (mainly for search after importing large size data)

**Reply by fgreco55** [reply]
Thanks.  This would be very useful info in your GettingStarted docs.
